### PR TITLE
DBZ-297 [WIP] Array type support for Postgres connector

### PR DESCRIPTION
**Pull request by emrul**
As suggested on Gitter, this is just a PR for feedback purposes of the initial support for array types in Postgres.

Note: Currently the PR does not handle composite types at all, only arrays!

### CODE REVIEW

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PgOid.java [hunk 1]
```diff
@@ -39,8 +39,11 @@ protected static int jdbcColumnToOid(Column column) {
         if (typeName.toUpperCase().equals("TSTZRANGE")) {
             return TSTZRANGE_OID;
         }
+        else if ( column.underlyingType() >= 0 ) {
```
Comment (rhauch) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PgOid.java]: What is the meaning of `0` here? Probably need a constant to make it clearer.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java [hunk 2]
```diff
@@ -86,11 +87,77 @@ public SchemaBuilder schemaBuilder(Column column) {
...
+            case PgOid.XML_ARRAY:
+                return null;
+            case PgOid.POINT_ARRAY:
+                return null;
+            case PgOid.JSONB_ARRAY:
+                return null;
+            case PgOid.JSON_ARRAY:
+                return null;
+            case PgOid.REF_CURSOR_ARRAY:
+                return null;
```
Comment (rhauch) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java]: Wouldn't it be a bit clearer if we used fallthrough here?

File: debezium-core/src/main/java/io/debezium/relational/ColumnEditor.java [hunk 3]
```diff
@@ -39,6 +39,14 @@
     int jdbcType();
 
     /**
+     * Get the underlying type for this column where
+     * jdbcType() itself is the generic Types.ARRAY.
+     *
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/relational/ColumnEditor.java]: Does this return the same thing as `jdbcType()` if the type is _not_ an array? Documenting the behavior would be helpful in the future.

File: debezium-core/src/main/java/io/debezium/relational/ColumnEditor.java [hunk 4]
```diff
@@ -140,6 +148,14 @@
     ColumnEditor jdbcType(int jdbcType);
 
     /**
+     * Set the underlying type for this column where
+     * jdbcType() itself is the generic Types.ARRAY.
+     *
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/relational/ColumnEditor.java]: Does this return the same thing as `jdbcType()` if the type is _not_ an array? Documenting the behavior would be helpful in the future.

File: debezium-core/src/main/java/io/debezium/relational/TableIdTransformerQuoted.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * Generic implementation of a TrableIdTransformer that accepts either
+ * a quote character to use for the start and end of every identifier
+ * part or two separate identifiers to use for start and end of each
+ * part.
+ *
+ */
+public class TableIdTransformerQuoted implements TableIdTransformer {
+
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/relational/TableIdTransformerQuoted.java]: The name of this class seems a bit strange. What about `TableIdQuoting`, `TableIdQuotingStrategy`, or something that better signifies what it does?

Also, does it make sense to add several static final instances for `DOUBLE_QUOTED`, `SINGLE_QUOTED`, and `SINGLE_AND_DOUBLE_QUOTED` instances? That would eliminate the need for [`PostgresTableIdTransformer`](https://github.com/debezium/debezium/pull/260/files#diff-a7160d27a9ce61bc7d146dda2cd65992), and they'd be more easily reusable.

Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/relational/TableIdTransformerQuoted.java]: Okay, I guess the name makes a bit more sense to me now. Perhaps the instances I mentioned above could go in `TableIdTransformer`??

File: debezium-core/src/main/java/io/debezium/relational/TableIdTransformerQuoted.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    @Override
+    public String toSqlQuoted(String part) {
+        if ( !part.startsWith(identifierStart) && !part.endsWith(identifierEnd)) {
+            return identifierStart + part + identifierEnd;
+        }
+        return part;
+    }
+}
```
Comment (rhauch) [inline debezium-core/src/main/java/io/debezium/relational/TableIdTransformerQuoted.java]: Would be really great to have good, thorough unit tests for this class.

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresSchemaIT.java [hunk 7]
```diff
@@ -88,6 +90,16 @@ public void shouldLoadSchemaForBuiltinPostgresTypes() throws Exception {
                               Schema.OPTIONAL_STRING_SCHEMA, Schema.OPTIONAL_STRING_SCHEMA);
             assertTableSchema("public.Quoted_Table", "Quoted_Text_Column",
                               Schema.OPTIONAL_STRING_SCHEMA);
+            assertTableSchema("public.array_table", "int_array, bigint_array, text_array",
+                              SchemaBuilder.array(Schema.OPTIONAL_INT32_SCHEMA).optional().build(), SchemaBuilder.array(Schema.OPTIONAL_INT64_SCHEMA).optional().build(),
+                              SchemaBuilder.array(Schema.OPTIONAL_STRING_SCHEMA).optional().build());
+            /*
+            assertTableSchema("public.composite_table", "c_type",
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresSchemaIT.java]: I'd suggest to focus on array support as of this PR and issue and handle composite types separately. That way we already can include and release the array support.

File: debezium-core/src/main/java/io/debezium/relational/Column.java [hunk 8]
```diff
@@ -49,6 +49,14 @@ public static ColumnEditor editor() {
     int jdbcType();
 
     /**
+     * Get the underlying type for this column where
+     * jdbcType() itself is the generic Types.ARRAY.
+     *
+     * @return a type constant for the specific database
+     */
+    int underlyingType();
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/Column.java]: Let's call it "component type", re-using the term established in the Java world for it?

File: debezium-core/src/main/java/io/debezium/relational/ColumnEditor.java (hunk 3 above)
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/ColumnEditor.java]: How about having it return null -- or even raise an `IllegalStateException` or so -- when invoking it for a non-array type?

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsStreamProducer.java [hunk 9]
```diff
@@ -478,11 +481,68 @@ protected Object extractValueFromMessage(PgProto.DatumMessage datumMessage) {
...
+                // representation (in Postgres) of the array value.
+                // The approach to decode this is sub-optimal but the only way to improve this is to update the plugin.
+                // Reasons for it being sub-optimal include:
+                // 1. It requires a Postgres JDBC connection to deserialize
+                // 2. The byte-array is a serialised string but we make the assumption its UTF-8 encoded (which it will
+                //    be in most cases)
+                // 3. For larger arrays and especially 64-bit integers and the like it is less efficient sending string
+                //    representations over the wire.
+                try {
+                    byte[] data = datumMessage.hasDatumBytes()? datumMessage.getDatumBytes().toByteArray() : null;
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsStreamProducer.java]: One more thing to consider is multi-dimensional arrays.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PgOid.java (hunk 1 above)
Comment (emrul) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PgOid.java]: It is just to set a default - I will go set it to a constant.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java (hunk 2 above)
Comment (emrul) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresValueConverter.java]: Yes, I will set all the ones I haven't implemented as fall throughs.

File: debezium-core/src/main/java/io/debezium/relational/ColumnEditor.java (hunk 3 above)
Comment (emrul) [inline debezium-core/src/main/java/io/debezium/relational/ColumnEditor.java]: given the return type is a privitive int we shouldn't really be returning a null.  I will document that it returns a default value when unset.

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresSchemaIT.java (hunk 7 above)
Comment (emrul) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresSchemaIT.java]: This makes sense, composite types are harder - @gunnarmorling can you split the issue for me?

File: debezium-core/src/main/java/io/debezium/relational/Column.java (hunk 8 above)
Comment (emrul) [inline debezium-core/src/main/java/io/debezium/relational/Column.java]: ok, will rename it.

File: debezium-core/src/main/java/io/debezium/relational/TableIdTransformerQuoted.java (hunk 5 above)
Comment (emrul) [inline debezium-core/src/main/java/io/debezium/relational/TableIdTransformerQuoted.java]: Ok, I have refactored this as you suggested.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsStreamProducer.java (hunk 9 above)
Comment (emrul) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsStreamProducer.java]: Yes, those are not currently supported by my work. I am not sure if the Postgres JDBC metadata returns the dimensionality of arrays (I would doubt it does) so we would probably need to issue another SQL query to get this info when refreshing schema.

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresSchemaIT.java (hunk 7 above)
Comment (gunnarmorling) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresSchemaIT.java]: Yep, I've extracted https://issues.jboss.org/browse/DBZ-314 and re-worded your original issue so it just refers to arrays.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsStreamProducer.java (hunk 9 above)
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsStreamProducer.java]: Ok, I take it you'd address that separately then? I've filed https://issues.jboss.org/browse/DBZ-315 as a follow-up. Help is welcome :)

Comment (emrul) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsStreamProducer.java]: @gunnarmorling Thank you for creating DBZ-315.  I plan to help there too but want to focus on getting my existing contributions into a state where they're fit to be incorporated and then I can start assessing how to proceed with the next steps.

**Comment by gunnarmorling** [comment]
Hum, it's a bit unfortunate that the last commit intermingles the DBZ-297 and DBZ-298. Makes it tough to grasp what's going on :(

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsSnapshotProducer.java [hunk 10]
```diff
@@ -148,7 +151,7 @@ private void takeSnapshot(Consumer<SourceRecord> consumer) {
             // we're locking in SHARE UPDATE EXCLUSIVE MODE to avoid concurrent schema changes while we're taking the snapshot
             // this does not prevent writes to the table, but prevents changes to the table's schema....
             schema.tables().forEach(tableId -> statements.append("LOCK TABLE ")
-                                                         .append(tableId.toString())
+                                                         .append(tableId.toQuotedId(TableIdTransformer.DOUBLE_QUOTED))
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsSnapshotProducer.java]: I think I've asked it before, but it still seems unaddressed: it seems we'll always quote the table name that way, correct? Shouldn't the transformer be instead be a part (member) of the table id which then would internally decide whether to apply it or not.

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresSchemaIT.java (hunk 7 above)
Comment (gunnarmorling) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresSchemaIT.java]: This comment should be removed now that complex types are left out of the picture.

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/RecordsStreamProducerIT.java [hunk 11]
```diff
@@ -92,8 +92,31 @@ public void shouldReceiveChangesForInsertsWithDifferentDataTypes() throws Except
...
+    }
+
+    @Test
+    public void shouldReceiveChangesForInsertsWithArrayTypes() throws Exception {
+        TestHelper.executeDDL("postgres_create_tables.ddl");
+
+        consumer = testConsumer(1);
+        recordsProducer.start(consumer);
+
+        // Quoted column name
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/RecordsStreamProducerIT.java]: Wrong comment?

File: debezium-connector-postgres/src/test/resources/postgres_create_tables.ddl [hunk 12]
```diff
@@ -9,3 +11,9 @@ CREATE TABLE time_table (pk SERIAL, ts TIMESTAMP, tz TIMESTAMPTZ, date DATE, ti
 CREATE TABLE text_table (pk SERIAL, j JSON, jb JSONB, x XML, u Uuid, PRIMARY KEY(pk));
 CREATE TABLE geom_table (pk SERIAL, p POINT, PRIMARY KEY(pk));
 CREATE TABLE tstzrange_table (pk serial, unbounded_exclusive_range tstzrange, bounded_inclusive_range tstzrange, PRIMARY KEY(pk));
+CREATE TABLE "Quoted_Table" (pk SERIAL, "Quoted_Text_Column" TEXT, PRIMARY KEY(pk));
+CREATE TABLE array_table (pk SERIAL, int_array INT[], bigint_array BIGINT[], text_array TEXT[], PRIMARY KEY(pk));
+
+-- CREATE TYPE composite_type_foo AS (f1 int, f2 text);
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/test/resources/postgres_create_tables.ddl]: These should be removed.

File: debezium-core/src/main/java/io/debezium/relational/TableIdTransformerQuoted.java (hunk 6 above)
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/TableIdTransformerQuoted.java]: Are you still planning to add unit tests for this?

Comment (emrul) [inline debezium-core/src/main/java/io/debezium/relational/TableIdTransformerQuoted.java]: No I wasn't planning to.  I'm always in favour of unit tests but this class seems to me small and very clear on its own.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsSnapshotProducer.java (hunk 10 above)
Comment (emrul) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/RecordsSnapshotProducer.java]: Sorry @gunnarmorling, I had just commented on this on DBZ-298 - https://github.com/debezium/debezium/pull/257#discussion_r129597318 - I agree, its unfortunate that the two commits got intermingled.  I did DBZ-298 and then branched DBZ-297 (which made sense at the time, I checked on Gitter before doing it).

File: debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java [hunk 13]
```diff
@@ -775,9 +775,7 @@ public void readSchema(Tables tables, String databaseCatalog, String schemaNameP
                             // ignore, some drivers don't have this index - e.g. Postgres
                         }
                         column.generated("YES".equalsIgnoreCase(autogenerated));
-                        if ( column.jdbcType() == Types.ARRAY) {
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/jdbc/JdbcConnection.java]: Why is it that you removed this check? It actually causes several test failures due to the assertion that makes sure that `componentType()` only is invoked for an array-typed column. I'll re-add it.

**Comment by gunnarmorling** [comment]
Ok, so I've extracted the bit s applying to array support, done some polishing, squashed it into one commit and applied it to master. I'll look into the quoting part next.

@emrul, could you send a PR for adding the newly supported column types to the docs (https://github.com/debezium/debezium.github.io/blob/develop/docs/connectors/postgresql.asciidoc)?

Thanks!

**Pull request closed without merging**